x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java

```diff
@@ -261,29 +421,108 @@ public static Hasher resolve(String name, Hasher defaultHasher) {
+            return Hasher.SSHA256;
+        } else {
+            throw new IllegalArgumentException("unknown hash format for hash [" + hashString + "]");
+        }
+    }
+
+    public static boolean verifyHash(SecureString data, char[] hash) {
+        try {
+            final Hasher hasher = resolveFromHash(hash);
+            return hasher.verify(data, hash);
+        } catch (IllegalArgumentException e) {
+            // The password hash format is invalid, we're unable to verify password
+            return false;
         }
-        return hasher;
+    }
+
+    private static char[] getPbkdf2Hash(SecureString data, int cost) {
+        try {
+            // Base64 string length : ((4*n/3)+3) rounded up to the next multiple of 4 because of padding = 44
+            CharBuffer result = CharBuffer.allocate(PBKDF2_PREFIX.length() + String.valueOf(cost).length() + 2 + 32 + 44);
+            result.put(PBKDF2_PREFIX);
+            result.put(String.valueOf(cost));
+            result.put("$");
+            char[] salt = SaltProvider.salt(32);  // 32 characters for 64 bytes
+            result.put(salt);
+            result.put("$");
+            SecretKeyFactory secretKeyFactory = SecretKeyFactory.getInstance("PBKDF2withHMACSHA512");
+            PBEKeySpec keySpec = new PBEKeySpec(data.getChars(), CharArrays.toUtf8Bytes(salt), cost, 256);
+            result.put(Base64.getEncoder().encodeToString(secretKeyFactory.generateSecret(keySpec).getEncoded()));
+            return result.array();
+        } catch (InvalidKeySpecException | NoSuchAlgorithmException e) {
+            throw new ElasticsearchException("Can't use PBKDF2 for password hashing", e);
+        }
+    }
+
+    private static boolean verifyPbkdf2Hash(SecureString data, char[] hash) {
+        try {
+            String hashStr = new String(hash);
+            if (!hashStr.startsWith(PBKDF2_PREFIX)) {
+                return false;
+            }
+            hashStr = hashStr.replace(PBKDF2_PREFIX, "");
+            String[] tokens = hashStr.split("\\$");
+            if (tokens.length != 3) {
+                return false;
+            }
+            int cost = Integer.parseInt(tokens[0]);
+            String salt = tokens[1];
+            SecretKeyFactory secretKeyFactory = SecretKeyFactory.getInstance("PBKDF2withHMACSHA512");
+            PBEKeySpec keySpec = new PBEKeySpec(data.getChars(), salt.getBytes(StandardCharsets.UTF_8), cost, 256);
+            String pwdHash = Base64.getEncoder().encodeToString(secretKeyFactory.generateSecret(keySpec).getEncoded());
+            return CharArrays.constantTimeEquals(pwdHash, tokens[2]);
+        } catch (InvalidKeySpecException | NoSuchAlgorithmException e) {
+            throw new ElasticsearchException("Can't use PBKDF2 for password hashing", e);
+        }
+    }
+
+    private static boolean verifyBcryptHash(SecureString text, char[] hash) {
+        String hashStr = new String(hash);
```

Comment:
Not new here but this is brought to light by this PR, we should avoid creating a new string from the hash. Instead we can use `CharArrays#charsBeginsWith`